item.lua

```diff
@@ -0,0 +1,340 @@
+--@module = true
+-----------------------------------------------------------
+-- helper functions
+-----------------------------------------------------------
+
+-- check whether an item is inside a burrow
+local function containsItem(burrow,item)
+    local res = false
+    local x,y,z = dfhack.items.getPosition(item)
+    if x then
+        res = dfhack.burrows.isAssignedTile(burrow, xyz2pos(x,y,z))
+    end
+    return res
+end
+
+-- fast reachability test for items that requires precomputed walkability groups for
+-- all citizens. Returns false for items w/o valid position (e.g., items in inventories).
+--- @param item item
+--- @param wgroups table<integer,boolean>
+--- @return boolean
+function fastReachable(item,wgroups)
+    local x, y, z = dfhack.items.getPosition(item)
+    if x then -- item has a valid position
+        local igroup = dfhack.maps.getWalkableGroup(xyz2pos(x, y, z))
+        return not not wgroups[igroup]
+    else
+        return false
+    end
+end
+
+--- @return table<integer,boolean>
+function citizenWalkabilityGroups()
+    local cgroups = {}
+    for _, unit in pairs(dfhack.units.getCitizens(true)) do
+        local wgroup = dfhack.maps.getWalkableGroup(unit.pos)
+        cgroups[wgroup] = true
+    end
+    cgroups[0] = false -- exclude unwalkable tiles
+    return cgroups
+end
+
+function table.contains (t, val)
+    for _,v in ipairs(t) do
+        if v == val then return true end
+    end
+    return false
+end
+
+
+
+-----------------------------------------------------------------------
+-- external API: helpers to assemble filters and `execute` to execute.
+-----------------------------------------------------------------------
+
+--- @alias conditions (fun(item:item):boolean)[]
+
+--- @param tab conditions
+--- @param burrow burrow
+--- @param outside boolean
+function condition_burrow(tab,burrow, outside)
+    if outside then
+        table.insert(tab, function (item) return not containsItem(burrow, item) end)
+    else
+        table.insert(tab, function (item) return containsItem(burrow, item) end)
+    end
+end
+
+--- @param tab conditions
+--- @param match number|string
+function condition_type(tab, match)
+    if type(match) == "string" then
+        table.insert(
+            tab,
+            function (item) return df.item_type[item:getType()] == string.upper(match) end
+        )
+    elseif type(match) == "number" then
+        table.insert(
+            tab,
+            function (item) return item:getType() == type end
+        )
+    else qerror("match argument must be string or number")
```

Comment:
`qerror` makes sense for when this is called from the commandline processing code, but if another tool is using the API, this should be `error` so problems will generate a stack trace